extensions/azurecore/src/account-provider/utils/simpleWebServer.ts

```diff
@@ -0,0 +1,100 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as http from 'http';
+import * as url from 'url';
+import { AddressInfo } from 'net';
+
+export type WebHandler = (req: http.IncomingMessage, reqUrl: url.UrlWithParsedQuery, res: http.ServerResponse) => void;
+
+export class SimpleWebServer {
+	private readonly pathMappings = new Map<string, WebHandler>();
+	private readonly server: http.Server;
+	private lastUsed: number;
+	private shutoffInterval: NodeJS.Timer;
+
+	constructor(private readonly autoShutoffTimer = 5 * 60 * 1000) { // Default to five minutes.
+		this.bumpLastUsed();
+		this.autoShutoff();
+		this.server = http.createServer((req, res) => {
+			this.bumpLastUsed();
+			const reqUrl = url.parse(req.url!, /* parseQueryString */ true);
+
+			const handler = this.pathMappings.get(reqUrl.pathname);
+			if (handler) {
+				return handler(req, reqUrl, res);
+			}
+
+			console.error('Unhandled request ', reqUrl);
+		});
+	}
+
+	private bumpLastUsed(): void {
+		this.lastUsed = new Date().getTime();
+	}
+
+	public async shutdown(): Promise<void> {
+		clearInterval(this.shutoffInterval);
+		return new Promise<void>((resolve, reject) => {
+			this.server.close((error) => {
+				if (error) {
+					reject(error);
+				} else {
+					resolve();
+				}
+			});
+		});
+	}
+
+	public startup(): Promise<string> {
+		let portTimeout: NodeJS.Timer;
+		const portPromise = new Promise<string>((resolve, reject) => {
+			portTimeout = setTimeout(() => {
+				reject(new Error('Timed out waiting for the server to start'));
+			}, 5000);
+
+			this.server.on('listening', () => {
+				// TODO: What are string addresses?
+				const address = this.server.address() as AddressInfo;
+				if (address!.port === undefined) {
+					reject(new Error('Port was not defined'));
+				}
+				resolve(address.port.toString());
+			});
+
+			this.server.on('error', () => {
+				reject(new Error('Server error'));
```

Comment:
This is kind of weird in that the port promise can reject after it resolves. And especially it gets interesting with error since an error might not necessarily mean that the server is shut down. Not sure how much of a concern it is though - seems pretty low risk. But maybe worth adding a simple check for to see if a port was already found.

Reply:
If that rejects later, it'll just end up in a blackhole as the resolve will have precedence. 

Eventually if we need an error listener we can setup the infrastructure around listening for errors, but rejecting after a resolve would do nothing from my understanding.

Reply:
Yeah I don't think there's much of any concern here. Although offhand I don't know what the behavior would be if something was to await on the Promise after it rejected - I think it would probably reject at that point? Very much an edge case though so you do whatever you think is appropriate here.